core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -482,9 +545,10 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
    *
    * '''NOTE''': This method must be invoked by the fiber itself.
    */
-  private[zio] def getChildren(): JavaSet[Fiber.Runtime[_, _]] = {
+  private def getChildren(): JavaSet[Fiber.Runtime[_, _]] = {
+    //executed by the fiber itself, no risk of racing with transferChildren
     if (_children eq null) {
-      _children = Platform.newConcurrentWeakSet[Fiber.Runtime[_, _]]()(Unsafe.unsafe)
+      _children = Platform.newWeakSet[Fiber.Runtime[_, _]]()(Unsafe.unsafe)
```

Comment:
There's a big optimization potential here. Currently we're initializing the Set using the default underlying size (can accommodate 12 entries before resizing).

Since now we can add children in bulk, we therefore know how many entries the set is meant to hold (at least initially). So we can instantiate it with a size large enough to accommodate all the entries without the need to dynamically resize.

Note that to calculate the size based on the number of expected entries you need to use this equation: `size = Math.ceil(nEntries / 0.75d)`

Reply:
I also thought about this potential, I think it goes beyond that since some of the spawned fibers ae very short lived and may already e finished by the time they're added to the fiber scope.
another thought I had, when forking multiple 'very similar' fibers (foreachParXXX), is it possible to come up with some kind of 'template' for the inherited fiber refs? they are all forked from the same parent so all should have very similar fiber refs except from the few which have a specialized 'fork' logic which perhaps can be pre-identified?

Reply:
btw, is it mandatory for `_children` to be a set? the `remove` operation is currently not used, we also never add the same fiber twice so the collection is effectively guaranteed to be a set.
I suspect an array based impl with 'occasional' GC can be sufficient here

Reply:
@kyri-petrou I suspect I may have to restore the synchronized weak set, seems like the set's `isEmpty` method effectively [modifies](https://github.com/openjdk/jdk/blob/jdk-21%2B35/src/java.base/share/classes/java/util/WeakHashMap.java#L371-L386) it. 

if there was a way to get the 'naive' set's size we'd still be ok, but given the current situation I think we must restore the synchronized weak set and seek alternative (I still think this doesn't have to be a set at all)